src/stats/dendrogram.jl

```diff
@@ -0,0 +1,178 @@
+# Node struct
+struct DNode{N}
+    idx::Int
+    position::Point{N, Float32}
+    children::Union{Tuple{Int,Int}, Nothing}
+end
+
+function DNode(idx::Int, point::Point{N}, children::Union{Tuple{Int,Int}, Nothing}) where N
+    return DNode{N}(idx, point, children)
+end
+
+"""
+    dendrogram(x, y; kwargs...)
+
+Draw a [dendrogram](https://en.wikipedia.org/wiki/Dendrogram),
+with leaf nodes specified by `x` and `y` coordinates,
+and parent nodes identified by `merges`.
+# Arguments
+- `x`: x positions of leaf nodes
+- `y`: y positions of leaf nodes (default = 0)
+# Keywords
+- `merges`: specifies connections between nodes (see below)
+- `treestyle`: one of `:tree`, `:box`.  Overload `dendrogram_connectors(::Val{:mystyle}, parent, child1, child2)` to define a new style.
+"""
+@recipe(Dendrogram, nodes) do scene
+    Theme(
+        weights = Makie.automatic,
+        branch_shape = :box,
+        linewidth = Makie.inherit(scene, :linewidth, 1.0),
+        color = Makie.inherit(scene, :color, :black),
+        colormap = Makie.inherit(scene, :colormap, :tableau_10),
+        colorrange = Makie.automatic,
+        orientation = :vertical,
+        groups = nothing,
+        cycle = [:color => :patchcolor],
+        inspectable = Makie.inherit(scene, :inspectable, false),
+        xautolimits = Makie.inherit(scene, :xautolimits, true),
+        yautolimits = Makie.inherit(scene, :yautolimits, true),
+    )
+end
+
+function recursive_dendrogram_points(node, nodes, ret_points = Point2f[], ret_colors = []; color=:black, branch_shape=:tree, groups=nothing)
+    isnothing(node.children) && return nothing
+    child1 = nodes[node.children[1]]
+    child2 = nodes[node.children[2]]
+   
+    l = dendrogram_connectors(Val(branch_shape), node, child1, child2)
+    
+    # even if the inputs are 2d, the outputs should be 3d - this is what `to_ndim` does.
+    append!(ret_points, Makie.to_ndim.(Point3f, l, 0))
+    push!(ret_points, Point3f(NaN)) # separate segments
+
+    if isnothing(groups)
+        cgroup = 0
+    else
+        gs = recursive_leaf_groups(node, nodes, groups)
+        @debug gs
+        cgroup = length(unique(gs)) == 1 ? first(gs) : 0
+    end
+
+    @debug cgroup maxlog=2
+
+    append!(ret_colors, [cgroup for _ in 1:length(l)])
+    push!(ret_colors, NaN32) # separate segments
+
+    recursive_dendrogram_points(child1, nodes, ret_points, ret_colors; branch_shape, groups)
+    recursive_dendrogram_points(child2, nodes, ret_points, ret_colors; branch_shape, groups)
+    return ret_points, ret_colors
+end
+
+
+function Makie.plot!(plot::Dendrogram{<: Tuple{<: Dict{<: Integer, <: Union{DNode{2}, DNode{3}}}}})
+    args = @extract plot (color, groups)
+
+    points_vec = Observable{Any}()
+    colors_vec = Observable{Any}()
+    
+    lift(plot[1], plot.branch_shape, plot[:color]) do nodes, branch_shape, color
```

Comment:
```suggestion
    points_vec = Observable{Vector{GeometryBasics.Point{2, Float32}}}([[0,0]])
    colors_vec = Observable{Any}([NaN])

    length(plot[1][])>1 && lift(plot[1], plot.branch_shape, plot[:color]) do nodes, branch_shape, color
```

this suggested change plots an empty graph should there be only 0 or 1 node.